### DISCUSSION TITLE
non-root installation of quadlet generator

### DISCUSSION THREAD

**Discussion by keturn** [containers/podman#23141] [Q&A]
While systemd allows users to add files to `~/.config/systemd/` to define new units, it does _not_ search user home directories for generators. (We can check this by running `systemd-path systemd-search-user-generator`.)

How can we define and run quadlet services without root?

**Comment by ygalblum** [comment]
Please see: https://docs.podman.io/en/latest/markdown/podman-systemd.unit.5.html#podman-rootless-unit-search-path for defining rootless Quadlets

**Reply by keturn** [reply]
That tells me where I can put quadlet unit files as a non-root user,  
but it does not tell me how to get a user-built instance of podman (with no write access to `/usr/lib/systemd/system-generators/`) to run the quadlet generator.

**Comment by keturn** [comment]
One workaround is to manually run the generator:

```sh
~/.local/lib/systemd/user-generators/podman-user-generator ~/.config/systemd/user/ && \
    systemctl --user daemon-reload
```

(Where that `podman-user-generator` is a symlink to `libexec/podman/quadlet`.)

It won't get run automatically, so you do need to remember to do it every time you change a unit file in `~/.config/containers/systemd/`

**Reply by ygalblum** [reply]
Quadlet and the soft links should be installed automatically. You should not need to do anything manually in order to get systemd to run Quadlet as a generator in both system and user levels. Which distribution and podman version are you using?

For example, I'm running Fedora 39 and podman 4.9.4 and the user generator link is set up by the RPM:
```
$ systemd-path systemd-search-user-generator
/run/systemd/user-generators:/etc/systemd/user-generators:/usr/local/lib/systemd/user-generators:/usr/lib/systemd/user-generators
$ ls -l /usr/lib/systemd/user-generators
total 40
lrwxrwxrwx. 1 root root    31 Mar 25 02:00 podman-user-generator -> ../../../libexec/podman/quadlet
-rwxr-xr-x. 1 root root 36824 May 28 03:00 systemd-xdg-autostart-generator
```

**Reply by Luap99** [reply]
The question is to set up the generator as rootless user, likely because the user does not have root to install anything.

Looking at https://www.freedesktop.org/software/systemd/man/latest/systemd.generator.html there is no user path listed there so that seems like a systemd question and noty quadlet.

**Reply by Luap99** [reply]
looks like this: https://github.com/systemd/systemd/issues/32423
